### PR TITLE
Add ECS run config and agent

### PR DESCRIPTION
**Pull request by jcrist**
This PR adds `RunConfig` support for flows running on ECS. It adds:

- A new `ECSRun` run config, for configuring flow runs on ECS
- A new `ECSAgent` for running `ECSRun` based flows
- A new CLI option `prefect agent ecs start` for starting the ECS agent

A new agent was used rather than modifying the existing fargate agent, as:

- The configuration options we wanted to expose at the agent level were completely different than those the fargate agent currently exposes. Maintaining backwards compatibility while transitioning to the new model would be difficult.
- Since the agent can deploy on either fargate or ec2, the "ECS" prefix makes more sense than the "fargate" prefix

Note that the ECS agent uses a `prefect agent <name> start` pattern instead of `prefect agent start <name>`. We plan to transition the other agents to this pattern as well (but haven't done so in this PR). Since this agent is new, it made sense to start with the new pattern from the get-go.

Like other run-config based flows, configuration of a flow run comes from two sources:

- Agent-side configuration, which configures defaults for all flow runs started by the agent
- Flow-side configuration, which provides options to override config values on a per-flow basis

## Agent side configuration

There are a few things a user might want to configure at the agent level:

### AWS credentials

AWS credentials are loaded using boto3's normal methods (environment variables or an `~/.aws/config` file. Given appropriate credentials, a new user can run:

```
prefect agent ecs start
```

with no other configuration, and things will "just work". Note that this *doesn't* include S3 access, so it will only work with flows using `Docker` storage. To enable S3 access, the user will also need to configure a `taskRoleArn` for the tasks (more on that later). I don't think we should do this for the user, but we should provide thorough documentation on how to go about creating and configuring this task role in our docs.

### Default parameters to pass to `register_task_definition`

If a user wants to override the default task definition template, they can pass in a path to a custom template via `--task-definition`. This takes a path to a yaml file containing any default options to pass to `register_task_definition` for a new flow. Note that the path could be local to the agent, or remote on e.g. `s3`. Prefect ships with a default template, so providing this isn't required.

### Default parameters to pass to `run_task`

If a user wants to override the default parametesr to pass to `run_task`, they can use the `--run-task-option`/`-r` flags, which take a `parameter=value` argument. Structured values can be passed in as json, and multiple options can be provided by passing duplicate parameters. For example:

```python
prefect agent ecs start -r overrides='{"taskRoleArn": "..."}'
```

**I'm not 100% happy with the ergonomics of this, so better suggestions welcome (perhaps these should be passed in via a file instead, like the task definition?).**

## Flow-side configuration

A user wanting to use run a flow on ECS needs to configure `flow.run_config` with an `ECSRun`. Like other run configs, an empty run-config will use all the agent-side defaults:

```python
from prefect.run_configs import ECSRun

# Use the agent-side defaults
flow.run_config = ECSRun()
```

The `ECSRun` class exposes two ways of overriding the task definition parameters configured on the agent:

- `task_definition`: A task definition spec to use (passed in as a dict). This will be stored in prefect cloud/server directly along with the flow.
- `task_definition_path`: A path to a task definition to use. If a local path, this will be loaded on init and stored as `task_definition` above (we accept local paths for ergonomics, so users won't have to do `open(...).read()` themselves). Remote paths can also be used (e.g. `s3://`), in which case they'll be loaded by the agent at runtime. This lets users who don't want to store a task definition in prefect cloud (for e.g. security reasons) use s3 instead.

For example:

```python
# Use a task definition stored on s3
flow.run_config = ECSRun(task_definition_path="s3://path/to/mydef.yaml")

# Use a task definition stored in memory
flow.run_config = ECSRun(task_definition={...})
```

If either option is provided, it will override the default task definition template configured on the agent via `--task-definition`.

For common things a user might want to override (`image`, `cpu`, `memory`, environment variables), we provide nice shorthand configuration options, which will be applied when registering/starting a task:

```python
flow.run_config = ECSRun(image="my-fancy-image", env={"HELLO": "WORLD"})
```

We also expose `run_task_kwargs`, which lets a user override any additional parameters to pass to `run_task` on startup. Note that we currently provide no way to store these outside of prefect cloud/server's database. IMO there's nothing "secretive" that could be passed in to `run_task` that couldn't also be configured on a task definition instead.

## Runtime Execution

When an ECS flow run is picked up by the agent, the following process occurs:

- The agent checks if an existing task definition has already been registered for this flow + version, based on tags stored on the task definition.
- If a task definition doesn't already exist, a new one is registered. The task definition is initialized from the template configured on either the `run_config` (if present) or the agent. Additional parameters required by prefect (environment variables, image, command, ...) are added to the template along with any options specified on the `run_config`.
- The parameters to `run_task` are inferred by merging defaults on the agent and additional options specified on the `run_config`.
- `run_task` is called with the parameters above and the task definition arn.

This scheme means that a new task definition is created for every flow version. *Note that if the task definition template changes on the agent, a flow that has already had a task definition registered for it won't see the changes on the agent until a new version of the flow is registered and run* (the agent wouldn't register a new task definition until then). Runtime options and environment variables passed in via `--env` will be applied to all task runs regardless. I think I'm ok with this (it's the simplest thing to implement). We could also workaround this by hashing the default template and storing a hash in a tag on the task definition to force re-registration on agent-side changes. This seems overly complex to me though.

Checklist:

- [x] adds new tests (if appropriate)
- [x] adds a change file in the `changes/` directory (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

### CODE REVIEW

**Comment by jcrist** [comment]
Registering the following flow:

```python
import os
from prefect import Flow, task
from prefect.run_configs import ECSRun
from prefect.environments.storage import S3
from prefect.utilities.logging import get_logger

@task
def echo_env(key):
    logger = get_logger()
    logger.info(f"Environment variable {key}={os.environ.get(key)}")

with Flow("test-ecs") as flow:
    echo_env("TESTING")

flow.run_config = ECSRun(
    image="jcrist/prefect-testing",
    env={"TESTING": "HELLO WORLD!"}
)
flow.storage = S3(bucket="jcristharif-testing")
flow.register(project_name="testing")
```

And running an agent with:

```
$ prefect agent ecs start --label s3-flow-storage -r overrides='{"taskRoleArn": "..."}'
```

leads to a successful execution for me:

```
28 October 2020,05:13:07 	agent	INFO	Submitted for execution: Task arn:aws:ecs:us-east-2:304383062342:task/default/2f11135f15f843eb99e5b3748ea2f8cc
28 October 2020,05:13:51 	prefect.S3	INFO	Downloading test-ecs/2020-10-28t22-11-04-045580-00-00 from jcristharif-testing
28 October 2020,05:13:51 	prefect.CloudFlowRunner	INFO	Beginning Flow run for 'test-ecs'
28 October 2020,05:13:52 	prefect.CloudTaskRunner	INFO	Task 'echo_env': Starting task run...
28 October 2020,05:13:52 	prefect	INFO	Environment variable TESTING=HELLO WORLD!
28 October 2020,05:13:53 	prefect.CloudTaskRunner	INFO	Task 'echo_env': Finished task run for task with final state: 'Success'
28 October 2020,05:13:53 	prefect.CloudFlowRunner	INFO	Flow run SUCCESS: all reference tasks succeeded
```

**Comment by jcrist** [comment]
My main question right now has to do with configuring kwargs to `run_tasks` on the agents. As is we expose everything via the `--run-task-option`/`-r` flag, but the syntax isn't the best IMO. A few comments/questions:

- We currently expose `--launch-type` and `--cluster` as top-level flags, since they're common large configuration switches. Are there other things a common user might need to set that we should elevate (perhaps `--task-role-arn`)? Or should we keep all other `run_task` kwargs equal behind a generic flag?

- We could also move `run_task` kwargs to file-based configuration like `--task-definition`. This would be nicer syntax-wise IMO, but would force a user to write a file if they need to tweak anything (which any user using s3 storage will for the `taskRoleArn`). Something like `--run-task-options options.yaml` (like `--task-definition`, this could also be a file stored on `s3`). Could go either way on this.

**Comment by jcrist** [comment]
> My main question right now has to do with configuring kwargs to run_tasks on the agents...

After some discussion with others:

- I'm moving to storing `run_task_kwargs` in a file, similar to the task definition. I think this is cleaner than trying to specify json on the commandline. Like `--task-definition`, the file passed to `--run-task-kwargs` can be stored remotely on S3 as well as locally.
- I've added `--task-role-arn` to the agent, to make specifying a default task role ARN easier. I've also added a `task_role_arn` top-level kwarg to `ECSRun`, for easier per-flow ARN configuration. This seems like a common enough field to configure that making it easier to do is worth it.

**Comment by jcrist** [comment]
Ok, this should be ready for review.

File: src/prefect/agent/ecs/agent.py [hunk 1]
```diff
@@ -0,0 +1,533 @@
...
+
+    out = deepcopy(opts1)
+
+    # Everything except 'overrides' merge directly
+    for k, v in opts2.items():
+        if k != "overrides":
+            out[k] = v
+
+    # Everything in `overrides` except `containerOverrides` merge directly
+    overrides = opts2.get("overrides", {})
```
[the remark this answers is not in the document]
Comment (madkinsz) [reply src/prefect/agent/ecs/agent.py:48]: Fair enough 👍

**Review by madkinsz — APPROVED** [review]

**Review by joshmeek — APPROVED** [review]
🚀

**Pull request merged**